sopt/src/main/scala/dagr/sopt/util/MarkDownProcessor.scala

```diff
@@ -0,0 +1,199 @@
+      val text = heading.getText.toString.trim
+      Chunk.unwrappable(indent, text, "=" * text.length, "")
+    case para : Paragraph =>
+      val lines = para.getChildren.flatMap(c => process(c, indent))
+      val text  = lines.mkString
+      Chunk.wrappable(indent, text, "")
+    case br: SoftLineBreak =>
+      if (br.getNext != null && br.getNext.isInstanceOf[SoftLineBreak]) Seq.empty else SpaceChunk
+    case list: BulletList =>
+      list.getChildren.flatMap(c => process(c, indent+1)).toSeq ++ Seq(Chunk.empty)
+    case item: BulletListItem =>
+      val children = item.getChildren.toSeq
+      val text     = process(children.head, indent)
+      val bullet   = Chunk.wrappable(indent, "* " + text.mkString).withGutter(2)
+      val subs     = children.tail.flatMap(c => withoutTrailingEmptyLine(process(c, indent + 1)))
+      bullet +: subs
+    case list: OrderedList =>
+      list.getChildren.zipWithIndex.flatMap { case (c, i) => process(c, indent+1, i+1) }.toSeq ++ Seq(Chunk.empty)
+    case item: OrderedListItem =>
+      val children = item.getChildren.toSeq
+      val text     = process(children.head, indent)
+      val prefix   = listPosition + ". "
+      val bullet   = Chunk.wrappable(indent, prefix + text.mkString).withGutter(prefix.length)
+      val subs     = children.tail.flatMap(c => withoutTrailingEmptyLine(process(c, indent + 1)))
+      bullet +: subs
+    case code : FencedCodeBlock =>
+      code.getContentChars.toString.lines.toSeq.flatMap(line => Chunk.unwrappable(indent+1, line)) ++ Seq(Chunk.empty)
+    case quote: BlockQuote =>
+      val lines = quote.getChildren.flatMap(c => process(c, indent))
+      val text  = lines.mkString
+      Seq(Chunk.wrappable(indent, text).withPrefix("> ")) ++ EmptyChunk
+    case link: Link =>
+      Seq(Chunk.wrappable(indent, s"${link.getText} (${link.getUrl})"))
+    case delimited: DelimitedNodeImpl =>
+      Seq(Chunk.wrappable(indent, delimited.getText.toString))
+    case text : Text =>
+      Seq(Chunk.wrappable(indent, node.getChars.toString))
+    case other =>
+      if (other.hasChildren) other.getChildren.flatMap(c => process(c, indent)).toSeq
+      else Seq(Chunk.wrappable(indent, other.getChars.toString))
+  }
+
+  /**
+      * Takes a chunk of text representing a paragraph, list item, line in a code block, etc. and
+      * formats it for output by:
+      *   - prepending the indenting space
+      *   - re-wrapping the text if required (e.g. code block lines don't get wrapped)
+      *   - applying any additional gutter (e.g. to follow-on lines in list items)
+      */
+  private def indentAndWrap(chunk: Chunk): Seq[String] = {
+    val indent = " " * (indentSize * chunk.indent)
+    val gutter = indent + (" " * chunk.gutter)
+
+    if (chunk.wrappable) {
+      val length = this.lineLength - indent.length - chunk.prefix.length
+      val lines  = new ListBuffer[String]()
+      val words  = chunk.text.split("\\s+").iterator.buffered
```

Comment:
I think that if we wanted to use terminal codes for formatting the way we'd do it is introduce them in the appropriate Chunks in `process`, and then here instead of just splitting on whitespace and using the length of the text, we could split on whitespace and calculate the length of each word minus any terminal codes, and use that for wrapping calculations.